message.go

```diff
@@ -0,0 +1,138 @@
+package ndp
+
+import (
+	"encoding"
+	"fmt"
+	"io"
+	"net"
+
+	"golang.org/x/net/icmp"
+	"golang.org/x/net/ipv6"
+)
+
+// A Message is a Neighbor Discovery Protocol message.
+type Message interface {
+	Type() ipv6.ICMPType
+	encoding.BinaryMarshaler
+	encoding.BinaryUnmarshaler
+}
+
+// MarshalMessage marshals a Message into its binary form and prepends an
+// ICMPv6 message with the correct type.
+//
+// It is assumed that the operating system or caller will calculate and place
+// the ICMPv6 checksum in the result.
+func MarshalMessage(m Message) ([]byte, error) {
+	mb, err := m.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+
+	im := icmp.Message{
+		Type: m.Type(),
+		// Always zero.
+		Code: 0,
+		// Calculated by caller or OS.
+		Checksum: 0,
+		// TODO(mdlayher): is this the correct MessageBody implementation?
```

Comment:
Looks like it might be? Remove the TODO?

Reply:
Yep, will do!